Rework plot architecture

boehmseb opened the pull request:
This PR reworks how plots can be generated with the tool suite.

## Plot Generators

`PlotGenerator`s are now responsible for creating one or more `Plots`.
They get instantiated with a general options, a plot config (as intended by se-sic/VaRA#545) and Plot(Generator) specific options.
Generators declare these options and the CLI for each generator is automatically generated at load-time.

## CLI

The plot driver discovers all plot generators and constructs CLI a command for each generator from its option declarations.
When executed, the created command instantiates the generator class with the general plot config object and the generator-specific options.
It then invokes the generator's `__call__()` function passing the remaining arguments that are common to all plots (e.g., the 'view' option).
The idea is that subclasses of PlotGenerator only instantiate the concrete plots with their specific options and, in general, should not need the common options or override the `__call__()` function.

As an example, the paper config overview plot is already compatible with the new architecture.

## Other stuff

- From now on, we use `click` as CLI library
- Artefacts have adapted their format to support plot generators
- Specific artefact types (plots/tables) are now managed in modules of their types (e.g., `plots.py` for plot artefacts) 

Resolves se-sic/VaRA#545
Related: se-sic/VaRA#628, se-sic/VaRA#605, se-sic/VaRA#574

boehmseb committed:
Draft plot generator CLI generator

boehmseb committed:
Sort requirements in setup.py

boehmseb committed:
New plot interface/CLI prototype.

PlotGenerators are responsible for creating one or more plots.
They get instantiated with a general plot config (as intended by se-sic/VaRA#545) and Plot(Generator) specific options.
Generators also declare these options as CLI options. Common options are provided as constants by the PlotGenerator class.
There is also a helper function to create such CLI option objects.
This separates the implementation detail of the CLI from the plot generation (at least if we further extend/improve this helper function).

The plot driver discovers all plot generators and constructs CLI a command for each generator from its option declarations.
When executed, the created command instantiates the generator class with the general plot config object (it is not yet decided/implemented how this is created) and the generator-specific options.
It then invokes the generator's `__call__()` function passing the remaining arguments that are common to all plots (e.g., the 'view' option).
The idea is that subclasses of PlotGenerator only instantiate the concrete plots with their specific options and, in general, should not need the common options or override the `__call__()` function.

As an example, the paper config overview plot is already compatible with the new architecture.

boehmseb commented on varats-core/varats/plot/plots.py:
```diff
@@ -180,3 +182,111 @@ def prepare_plots(**args: tp.Any) -> tp.Iterable['varats.plot.plot.Plot']:
+    PLOT: tp.Type['varats.plot.plot.Plot']
+    OPTIONS: tp.List[CLIOptionTy]
+
+    def __init__(self, plot_config: PlotConfig, **plot_kwargs: tp.Any):
+        self.__plot_config = plot_config
+        self.__plot_kwargs = plot_kwargs
```
I would like to not store the kwargs here at all. Instead, each concrete generator should know what it needs (a similar thing could be done now for plots, too). The only reason I need this here is so that the constructor signatures match.
Any ideas/opinions on how to do this the best way?

boehmseb commented on varats-core/varats/plot/plots.py:
```diff
@@ -180,3 +182,111 @@ def prepare_plots(**args: tp.Any) -> tp.Iterable['varats.plot.plot.Plot']:
+    def generate(self) -> tp.List['varats.plot.plot.Plot']:
+        """This function is called to generate the plot instance(s)."""
+
+    def __call__(self, common_options: CommonPlotOptions):
```
The idea is that this function handles everything that is common for all plots, i.e., the parameters for the `vara-plot` command and _not_ its subcommands (aka the concrete plot generators). The reason for this is, that those options should to the same thing every time. Otherwise, they should be plot-specific.

boehmseb commented on varats-core/varats/utils/cli_util.py:
```diff
@@ -80,3 +81,22 @@ def initialize_logger_config() -> None:
     logging.basicConfig(level=log_level)
+
+
+CLIOptionTy = tp.Callable[..., tp.Any]
```
Unfortunately, click is completely untyped :frowning_face:

boehmseb commented on varats-core/varats/utils/cli_util.py:
```diff
@@ -80,3 +81,22 @@ def initialize_logger_config() -> None:
+
+
+# TODO: make this have a typed version of the click.core.Option constructor?
+def make_cli_option(*param_decls: str, **attrs: tp.Any) -> CLIOptionTy:
```
We could use this function to hide the concrete CLI library we are using, or at least provide a typed version. Opinions?

boehmseb committed:
Simplify usage of `@check_required_args()`

boehmseb committed:
Add click dependency

boehmseb committed:
Move common and plot config options declaration to their classes.

boehmseb committed:
Move plots and tables modules to from core to varats

boehmseb committed:
Add common plot option --file-type

boehmseb committed:
Remove old plot driver code

boehmseb committed:
Remove old plot generation code

boehmseb committed:
Remove PlotRegistry, use __init_subclass__() instead.

cormensratio committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Replaces '-v True' with '-v' to show plot.

cormensratio committed:
Modifies PlotGenerator CLI options.

cormensratio committed:
Adds first sankey plot generator implementation.

vulder committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Merge branch 'f-PlotRework' of github.com:se-sic/VaRA-Tool-Suite into f-PlotRework

cormensratio committed:
Adds optional revision + WIP: Adapts sankey plot to new generator

cormensratio committed:
Remove revision from required args in sankey plot

cormensratio committed:
Adds optional case study option.

boehmseb committed:
Make view a flag again

boehmseb committed:
Make artefacts compatible with new plot architecture (WIP)

boehmseb committed:
Fix some type errors

cormensratio committed:
Adds types to sankey generator attributes

cormensratio commented on varats/varats/plot/plot.py:
```diff
@@ -17,14 +17,63 @@ class PlotDataEmpty(Exception):
         self.__saved_extra_args = kwargs
 
+    @classmethod
+    def __init_subclass__(cls, plot_name: str, **kwargs: tp.Any) -> None:
```
I think that the plot discovery currently fails because the signature doesn't match. At least my IDE highlights this ;).

cormensratio committed:
Adapt sankey plot params to cli options

cormensratio committed:
Moves cli options to top of module

cormensratio committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Adds graphviz cli options

cormensratio committed:
Adds file type choices

cormensratio committed:
Adds more cli choices

boehmseb committed:
Upgrade to click 8.0.1

boehmseb committed:
Move cli_util.py from core to varats

boehmseb committed:
Add EnumChoice parameter type

boehmseb committed:
Prompt when trying to view mutliple plots at once

boehmseb committed:
Add exception that can be thrown during plot generator initialization.

cormensratio committed:
Change misbehaving click.Path path_type to pathlib.Path

cormensratio committed:
Pass plot_name to plots as class kwarg to fulfill __init_subclass__ requirement.

cormensratio committed:
Replaces enum hack in click.Choice with EnumType

cormensratio committed:
Adapts sankey-plot to cli interface for single plot generation of a specific revision

cormensratio committed:
Remove unreachable code

boehmseb committed:
CLI EnumType support all click.Choice parameters

boehmseb committed:
Plot: do not register subclass if plot_name is None

boehmseb committed:
Fix __init_subclass__() docs for Plot

cormensratio committed:
Revert "Change misbehaving click.Path path_type to pathlib.Path"

This reverts commit 4805540ace05629cdf9a1046916341aca44e00b1.

cormensratio committed:
Adapts sankey plot to require a CaseStudy

cormensratio committed:
Adds sankey plot generators for a revision, case studies, and paper configs

cormensratio committed:
Adds graphviz plot generators for a revision, case study, and paper config

cormensratio committed:
Deduplicate case study filename validation

boehmseb committed:
Fix incorrect generator class being chosen for plot subcommands.

boehmseb commented on varats/varats/paper_mgmt/paper_config.py:
```diff
@@ -91,6 +91,27 @@ def get_all_case_studies(self) -> tp.List[CaseStudy]:
             for case_study in case_study_list
         ]
 
+    def get_all_case_study_filenames(self) -> tp.List[str]:
+        """
+        Generate a list of all case study file names in the paper config.
+
+        Returns:
+            list of all case study file names in the paper config
+        """
+        filenames: tp.List[str]
+        _, _, filenames = next(os.walk(self.__path))
+        return list(filter(lambda a: a.endswith(".case_study"), filenames))
+
+    def get_all_case_study_paths(self) -> tp.List[Path]:
+        """
+        Generate a list of all paths to case studies in the paper config.
+
+        Returns:
+            list of all case study paths in the paper config
+        """
+        filenames: tp.List[str] = self.get_all_case_study_filenames()
+        return [self.__path / filename for filename in filenames]
+
```
We can simply use `PaperConfig.get_all_case_studies()` instead. No need to manually collect those files.

cormensratio replied:
Right, we can gather all case studies of the current PC with `PaperConfig.get_all_case_studies()`, but I thought about a way to let the user specify which case study they want to select for the current plot by choosing a unique name of the case study in the cli. Since the user cannot pass a CaseStudy object to the cli, I used the case study file name as identifier. Furthermore, we could use this to show the user a list of available case study files, if an incorrect one is selected. Even better would be a `click.Choice` that already provides available case studies, but this is currently not possible because of a cyclic dependency that we already discussed last time.
What other unique 'key' of a case study besides its file name do we have that a user can pass to the cli in order to specify the case study?

boehmseb replied:
The only unique identifier is name + version. Maybe I manage to finish my Choice implementation of this in the evening, than this problem would be gone for good.

boehmseb commented on varats/varats/plots/blame_interaction_degree.py:
```diff
@@ -844,22 +954,48 @@ def _build_graphviz_fig(
 
 
-class BlameLibraryInteraction(Plot):
+def _validate_cs_filename(cs_file_name: str) -> CaseStudy:
+    """
+    Checks if the filename of the selected case study exists in the current
+    paper config.
+
+    Args:
+        The file name string of the case study
+
+    Returns:
+        the case study
+    """
+
+    paper_conf: PC.PaperConfig = PC.get_paper_config()
+    if cs_file_name not in paper_conf.get_all_case_study_filenames():
+        all_cs_string = [
+            cs + " " for cs in paper_conf.get_all_case_study_filenames()
+        ]
+        raise FileNotFoundError(
+            f"The selected case study filename could not be found in the "
+            f"current paper config. Did you mean?\n {all_cs_string}"
+        )
+    return PC.load_case_study_from_file(paper_conf.path / cs_file_name)
```
This should be in the `case_study` module. 
Should be obsolete anyway (see my comment in the `paper_config` module.

boehmseb commented on varats/varats/plots/blame_interaction_degree.py:
```diff
@@ -1474,7 +1533,126 @@ def calc_missing_revisions(self, boundary_gradient: float) -> tp.Set[str]:
+        ]
+
+
+class SankeyLibraryInteractionsGeneratorCS(
```
We have multiple options on how to handle the a situation "generate a plot for each case study in my paper config":

- do it as @cormensratio implemented it with multiple different generators
- we could merge this generator and the PC version simply by making the case study optional and selecting the behavior depending on whether a case study is selected
- the PC case is a common pattern that matches many plots and until now is handled by the artefacts mechanism:
  you can currently specify an artefact with an option `paper_config=True` and the artefact toolchain will call the plot (or the generator in the new architecture) once for each case study in the current paper config and no plot(generator) should implement such a paper config behavior. This means that such a behavior can only be achieved using artefacts.

I personally like the last option since it "just works" for all plots that require a (mandatory or optional) case study (this can also be checked with our new architecture). This also means less or less cluttered plot generators and one can argue that if you are at the point that you need to generate the same plot you should use artefacts anyway.

@vulder what's your view on this?

vulder replied:
Having everything unified with good reuse sounds not to bad, however, I'm not 100% sure about always requiring artefacts. From what I found, listing to users, people already have problems with the paper config/case study model when they start working with varats. For example, they struggle why BB is not executing things or why suddenly so many different revisions are executed, how to get an overview of everything, and how to create/manipulate case studies.
I'm currently a bit torn about this. From one point of view, things could become easier with "How to" guides in the documentation, from another point of view, a lot of our tooling needs a bit of polishing to make things like: "I just want on revision of `foo` as a case study" easier. (Currently, one has to use the case study generator, pic a distribution, set the sample size to 0, and add extra revs, which is the rev one wants... also having to look up the rev in the git.)
Now, we want to further hardly tie artefacts together with plots, which on it self does not have to be a bad thing, but forces users to build up/understand artefacts and artefact managment "just" to get their 3 examples plots.

TDL: In general, I agree that making things reusable is probably the best route design wise, I just want to remark that we need to work and improve on the usability side, which currently is lacking and could get even worse.

boehmseb replied:
I see your point. 
However, solution would not force a user to always use artefacts. Only in the case that a user already has multiple case studies and she wants to generate the same plot for all of them.

I also just had an idea for a fourth possibility:
We currently plan to enhance the case study command line option to properly check whether the user gives a valid case study. We could also add a special value to that option that means "use all case studies". This would force the implementer to handle this case and thus, give the same guarantees as the artefacts solution. This would also simplify the artefacts code a little bit since this case needs no special handling anymore but it would be a breaking change to the artefacts file format (however, this could happen during this PR anyway).

boehmseb commented on varats/varats/plots/blame_interaction_degree.py:
```diff
@@ -50,6 +56,120 @@ class EdgeWeightThreshold(Enum):
+
+# TODO: Remove param default values in plot generation if already set in cli
+
+OPTIONAL_FIG_TITLE: CLIOptionTy = make_cli_option(
```
Some of these options will move to the plot config object at some point, so we should make sure they are easy to migrate.

cormensratio committed:
Changes paper_config_overview_plot generator name to 'pc-overview-plot'

cormensratio committed:
Adapts the 'case_study_overview_plot' to generators.

cormensratio committed:
Adds docstring for cs overview plot

boehmseb committed:
Implement case study choice parameter.

boehmseb committed:
Implement dry-run option for vara-plot

boehmseb committed:
Decouple artefacts and plots/tables.

This gets rid of that nasty import cycle once and for all by moving concrete artefact implementations to the plots and tables modules and installing our usual __init_subclass__() based register mechanism so that the artefacts module does not need to import anymore. The discovery of artefact implementations can be triggered if needed using the initialize_artefacts() method but usually, those the relevant modules have already been imported somewhere, so initialization will most of the time happen "by accident".

boehmseb committed:
Move TableFormat to tables module to avoid import cycles.

This would be necessary anyways when we migrate tables to the same new architecture we currently implement for plots.

boehmseb committed:
Cleanup artefact refactoring

boehmseb committed:
Fix some type issues

boehmseb committed:
Fix tests

boehmseb committed:
Fix EnumChoice name

boehmseb committed:
Merge branch 'vara-dev' into f-PlotRework

boehmseb committed:
Fix broken import

boehmseb committed:
Fix crash when no paper config is set

boehmseb committed:
Fix integration test

boehmseb committed:
Fix duplicate key

boehmseb committed:
Fix another duplicate key

boehmseb committed:
That did not quite fix it :(

boehmseb committed:
Add missing mkdir

boehmseb committed:
Fix plot artefact output path

boehmseb committed:
Remove empty file

vulder commented:
related se-sic/VaRA#628

cormensratio committed:
Merge branch 'f-PlotRework' of github.com:se-sic/VaRA-Tool-Suite into f-PlotRework

cormensratio committed:
Adds blame interaction degree generator for a specific revision

cormensratio committed:
Fix interaction degree generator

cormensratio committed:
Adds type information to cli options

cormensratio committed:
Adds default legend_title and fig_title for interaction-degree-plot

cormensratio committed:
Fix docstring

cormensratio committed:
Adds TODO

boehmseb commented on varats/varats/plots/bug_annotation.py:
```diff
@@ -22,21 +20,8 @@ def draw_bugs(
-            * color
-            * vertical_alignment
-            * label_size
+        plot_kwargs: the arguments that specify a plots style
```
I don't like the idea of relying on kwargs here since this function could be called from any  plot.
Since it always takes the same couple of arguments, just make them explicit.
I also suggest to pass a `PlotConfig` object as an additional argument since some of the required arguments will definitely be migrated to the plot config.

Same goes for `draw_cves()` btw.

boehmseb commented on varats/varats/plots/cve_annotation.py:
```diff
@@ -22,21 +20,8 @@ def draw_cves(
-            * color
-            * vertical_alignment
-            * label_size
+        plot_kwargs: the arguments that specify a plots style
```
See comment on `draw_bugs()`

boehmseb commented on varats/varats/ts_utils/cli_util.py:
```diff
@@ -84,19 +86,114 @@ def initialize_logger_config() -> None:
+        choices: tp.Dict[str, tp.List[ChoiceTy]],
+        case_sensitive: bool = True
+    ):
+        # Relates to: https://thingspython.wordpress.com/2010/09/27/another
+        # -super-wrinkle-raising-typeerror/
+        self.as_super = super(TypedMultiChoice, self)
```
I'm not sure if that is really what happens. The linked article is for Python 2.4/2.6.

Could you give me a MWE to reproduce the issue so I can confirm this?

boehmseb committed:
Merge branch 'vara-dev' into f-PlotRework

boehmseb committed:
Fix tests

boehmseb committed:
Fix some problems caused by merge

boehmseb committed:
Initialize plot config from params

boehmseb committed:
Fix extra tests

cormensratio committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Adds missing commit_map

cormensratio committed:
Adds missing comparator for commit ambiguity check

boehmseb committed:
Make -h work with vara-plot

boehmseb committed:
Convert vara-art to click + refactoring

`vara-art add` was removed in favour of a new option `--save-artefact` to `vara-plot` (same needed for `vara-table`)

cormensratio committed:
Adapts evaluation database and blame_interaction_degree plot to the use of Short-/FullCommitHashes to replace revision strings

cormensratio committed:
Store revisions as strings

cormensratio committed:
Adds todos

cormensratio committed:
Fix test

cormensratio committed:
Update varats/varats/plots/repository_churn.py

Co-authored-by: Sebastian Böhm <boehmseb@cs.uni-saarland.de>

cormensratio committed:
Apply suggestions

cormensratio committed:
Fix x-ticks for code churn

cormensratio committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Merge f-ReplacesRevStringsWithCommitHashType into f-PlotRework

boehmseb committed:
Make EnumChoice accept Enums into its convert function

cormensratio committed:
Fix inconsistencies

cormensratio committed:
Call set_xticks() to suppress matplotlib UserWarning.

cormensratio committed:
Merge branch 'f-ReplacesRevStringsWithCommitHashType' into f-PlotRework

boehmseb committed:
Fix possible issue with super() in Typed[Multi]Choice

cormensratio committed:
Removes redundant lambda

cormensratio committed:
Merge branch 'f-PlotRework' of github.com:se-sic/VaRA-Tool-Suite into f-PlotRework

cormensratio committed:
Merge branch 'f-ReplacesRevStringsWithCommitHashType' into f-PlotRework

vulder committed:
Merge branch 'vara-dev' into f-ReplacesRevStringsWithCommitHashType

boehmseb committed:
Merge branch 'vara-dev' into f-PlotRework

boehmseb committed:
Fix broken import

boehmseb committed:
Test custom click choice types

boehmseb committed:
More tests for artefacts

boehmseb committed:
Test driver_plot

boehmseb committed:
Fix extra test

boehmseb committed:
Fix some warnings

cormensratio committed:
Moves click options to PlotConfig + Adapts sankey-, graphviz-, interaction-degree-, and case-study-overview-plot to PlotConfig and CommitHash types

cormensratio committed:
Merge branch 'f-ReplacesRevStringsWithCommitHashType' into f-PlotRework

cormensratio committed:
Merge branch 'f-PlotRework' of github.com:se-sic/VaRA-Tool-Suite into f-PlotRework

cormensratio committed:
Adds missing commit hash completion

cormensratio committed:
Adapt _calc_missing_revisions() to cli

cormensratio committed:
Formatting

cormensratio committed:
Small (pylint-)fixes

boehmseb committed:
Add plotting guide

boehmseb committed:
Fix some docs-related warnings

boehmseb committed:
Add docs for plotting internals

boehmseb committed:
Remove PlotGenerator.PLOT

This wasn't really used anywhere and removing this allows generators to generate plots of different types if they really want to.

boehmseb committed:
Make matplotlib style a plot config option

cormensratio committed:
Move dpi to plotConfig

cormensratio committed:
Merge branch 'f-PlotRework' of github.com:se-sic/VaRA-Tool-Suite into f-PlotRework

cormensratio committed:
Adds missing plot_config parameter

cormensratio committed:
Removes unused parameter

cormensratio committed:
Some fixes

cormensratio committed:
Add report type shorthands as click.Choice() to cli

cormensratio commented on varats/varats/plots/paper_config_overview.py:
```diff
@@ -286,6 +289,8 @@ def calc_missing_revisions(
         revisions = _gen_overview_plot_for_project(**self.plot_kwargs)
         revisions.sort_values(by=['revision'], inplace=True)
+
+        # TODO: Adapt to click cli.
```
Which `case_study`(ies) do we use to get the correct `commit_map`s ?

cormensratio commented on varats/varats/plot/plots.py:
```diff
@@ -0,0 +1,559 @@
+    )
+    REQUIRE_REPORT_TYPE: CLIOptionTy = make_cli_option(
+        "--report-type",
+        type=click.Choice([
```
Should we directly convert the report `SHORTHAND` strings to their corresponding `report_type` when saving them in the generators as attributes? 
E.g.:
`self.__report_type: BlameReport = BaseReport.lookup_report_type_by_shorthand(plot_kwargs["report_type"])`

boehmseb replied:
I changed that to use the full report name. I think that's better readable for users.

cormensratio commented on docs/source/tutorials/visualizing_data.rst:
```diff
@@ -0,0 +1,63 @@
+.. program-output:: vara-plot -h
+    :nostderr:
+
+To see what plot-specific a plot has, you need to look at the help output for that specific plot, e.g.:
```
```suggestion
To see what plot-specific options a plot has, you need to look at the help output for that specific plot, e.g.:
```

cormensratio commented on docs/source/tutorials/visualizing_data.rst:
```diff
@@ -0,0 +1,63 @@
+Generating Plots and Tables
+---------------------------
+
+Plots and tables can be generated with the tools :ref:`vara-plot` and :ref:`vara-table`.
```
I think that the image from `:ref: here <Plots>` under `Plot Architecture` would be (also) relevant to see on this page, since it gives a nice overview of how we can call a plot and is not only relevant `If you want to write your own plots ...`.

cormensratio commented on docs/source/tutorials/visualizing_data.rst:
```diff
@@ -0,0 +1,63 @@
+    :nostderr:
+
+The plot files are stored in ``$VARATS_ROOT/plots``.
+You can group plot files into subdirectories using the option ``--plot-dir``, e.g. ``vara-plot --plot-dir=foo pc-overview-plot --report-type=EmptyReport`` will create a file in the directory ``$VARATS_ROOT/plots/foo``.
```
Has to be changed if we decide to use the reports' `SHORTHAND` as click.Choices.
I don't think we need the `=`?
```suggestion
You can group plot files into subdirectories using the option ``--plot-dir``, e.g. ``vara-plot --plot-dir foo pc-overview-plot --report-type EMPTY`` will create a file in the directory ``$VARATS_ROOT/plots/foo``.
```

boehmseb replied:
The `=` is optional but typically used with long-style options.
As for the reports, I'm not sure whether it is good to use the shorthands, since they are less self-documenting than the full report names. @vulder wha's your opinion on this?

vulder replied:
I agree, the `SHORTHANDS` are often not very readable (e.g., `BlameReport` vs `BR`). If people want it, we would add a lookup mechanic to allow both, but from my point of view, the default style we advertise should be the report name.

cormensratio commented on docs/source/tutorials/visualizing_data.rst:
```diff
@@ -0,0 +1,63 @@
+For example, an artefact that will generate a ``pc-overview-plot`` for the current paper config can be added
+via::
+
+    vara-plot --save-artefact="PC Overview" pc-overview-plot --report-type=EmptyReport
```
s.o.
```suggestion
    vara-plot --save-artefact "PC Overview" pc-overview-plot --report-type EMPTY
```

cormensratio replied:
Do we need the `=`?
```suggestion
    vara-plot --save-artefact "PC Overview" pc-overview-plot --report-type=EmptyReport
```

cormensratio commented on docs/source/vara-ts-api/plot_architecture.svg:
```diff
@@ -0,0 +1,492 @@
+         id="tspan848"
+         x="2.6458333"
+         y="5.2916665"
+         style="font-style:normal;font-variant:normal;font-weight:normal;font-stretch:normal;font-size:3.52778px;font-family:'DejaVu Sans Mono';-inkscape-font-specification:'DejaVu Sans Mono';stroke-width:0.264583">vara-plot --plot-dir=foo --title=Overview pc-overview-plot --report-type=EmptyReport</tspan></text>
```
If the colors do not decrease the text's readability, maybe color the parts of the command line string according to the underlying parentheses' color?

boehmseb replied:
I mainly didn't do it because inkscape didn't let me use different colors in the same text box and I was too lazy to split it into multiple text boxes so that they still align properly.

boehmseb replied:
I tried this but in my opinion, the black text is better readable.

cormensratio commented on docs/source/vara-ts-api/plot_architecture.svg:
```diff
@@ -0,0 +1,492 @@
+         id="tspan924-47-3-9"
+         x="134.9375"
+         y="58.208336"
+         style="font-size:3.52778px;fill:#000000;fill-opacity:1;stroke-width:0.264583">Plot</tspan></text>
```
Maybe this is too captious, but are the three dots under the last Plot box horizontally centered? ;)

cormensratio commented on docs/source/vara-ts-api/plots.rst:
```diff
@@ -3,50 +3,81 @@ Plots
+4. *Plot- or table-specific options*: options specific to a certain plot
+
+Plot generation is handeled by the :class:`~varats.plot.plots.PlotGenerator` classes.
+The CLI selects a concrete plot generater via the passed **plot generator name** and immeadiately calls its ``__call__()`` function (which is implemented by the :class:`~varats.plot.plots.PlotGenerator` itself and must not be overridden by subclasses).
```
```suggestion
The CLI selects a concrete plot generator via the passed **plot generator name** and immediately calls its ``__call__()`` function (which is implemented by the :class:`~varats.plot.plots.PlotGenerator` itself and must not be overridden by subclasses).
```

boehmseb committed:
Fix some typos

boehmseb committed:
Fix tests

boehmseb committed:
Apply suggestions from review.

cormensratio committed:
Swap calc_missing_revisions() of paper_config_overview with case_study_overview

boehmseb committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Fix mypy

cormensratio committed:
Fix pylint

cormensratio committed:
Deletes excluded plots 'b_cluster_analysis', 'interaction_graph' and their corresponding modules 'blame_diff_cluster_analysis.py' and 'commit_interactions.py' as they were the only plots in those modules.

cormensratio committed:
Adds todo

boehmseb committed:
Implement metaclass-based default value handling for plot config options.

Needs to be changed because mypy and IDEs cannot handle the generated functions.

boehmseb committed:
Implement less fancy default value handling for plot config options that works with static typing.

Downsides: it requires more typing and duplication.

boehmseb commented:
Ready for a first round of review.
@vulder: focus on `driver_plot.py`, `plots.py` and `plot.py`.
There is also some [documentation](https://vara.readthedocs.io/en/f-plotrework/tutorials/visualizing_data.html) available.
For a simple example plot take a look at `paper_config_overview.py`.

boehmseb committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio commented on varats/varats/plot/plots.py:
```diff
@@ -0,0 +1,686 @@
+                help=self.__help
+            )
+
+    def value_or_default(self, default: tp.Optional[OptionType]) -> OptionType:
```
```suggestion
    def value_or_default(self, default: tp.Optional[OptionType] = None) -> OptionType:
```

cormensratio commented on varats/varats/plots/blame_interaction_degree.py:
```diff
@@ -929,42 +1047,23 @@ def _get_degree_data(self) -> pd.DataFrame:
 
-        style.use(self.style)
-        commit_map: CommitMap = self.plot_kwargs['get_cmap']()
+        style.use(self.plot_config.style())
```
Calling `self.plot_config.style()` returns the unset `value`, which is `None` instead of the wanted `default`.

boehmseb replied:
How so?

boehmseb replied:
Nevermind. That happens when you forget a return :man_facepalming:

cormensratio replied:
:D

boehmseb committed:
Remove unused function

boehmseb committed:
Add missing default parameter

boehmseb committed:
Add missing return

boehmseb committed:
Add missing ()

boehmseb committed:
Fix type errors

boehmseb committed:
Fix some pylint issues

boehmseb committed:
Add plot config tests

boehmseb committed:
Use _ instead of - in plot config option names.

This makes it more consistent with the accessor names and we can simply do the substitution when generating the CLI options to get nicer command line names.

boehmseb committed:
Add missing docstring

boehmseb committed:
Fix tests

cormensratio committed:
Changes metavars of required cli options

cormensratio committed:
Removes resolved TODO

cormensratio committed:
Adds missing parentheses to plot_config.style

cormensratio commented on varats/varats/plot/plots.py:
```diff
@@ -0,0 +1,690 @@
+                ),
+                PlotConfigOption(
+                    "show_legend",
+                    default=False,
```
Maybe change this to True?
```suggestion
                    default=True,
```

boehmseb replied:
Hmm, I'd say that it is better if a plot needs to opt in to having a legend (i.e., by using `PlotConfig.show_legend(True)`)

cormensratio replied:
Right, this would be better.

cormensratio committed:
Decouple bug/cve annotation styles + Makes bug/cve parameters explicit

cormensratio reviewed (approved):
Looks good. ~~Maybe consider setting the `show_legend` option `default` to `True`.~~

boehmseb committed:
cleanup: artefacts

boehmseb committed:
cleanup: artefacts part II

boehmseb committed:
cleanup: plot.py

boehmseb committed:
cleanup: plots.py

boehmseb committed:
move CLI option declarations to top level

boehmseb committed:
Fix missing assignment in PlotConfig constructor

boehmseb committed:
Add missing module docstring.

cormensratio committed:
Cleanup interaction-degree-plot

cormensratio committed:
Cleanup sankey-plot-rev and sankey-plot-cs

cormensratio committed:
Fix line-too-long

cormensratio committed:
Fix mypy

cormensratio committed:
Remove generator attributes

cormensratio committed:
Merge branch 'vara-dev' into f-PlotRework

cormensratio committed:
Remove unused import

vulder commented on docs/source/tutorials/visualizing_data.rst:
```diff
@@ -0,0 +1,65 @@
+
+.. note::
+
+    In the following, we will only write about plots since plots and tables work essentially in the same way.
```
```suggestion
    In the following, we only write about plots since plots and tables work essentially in the same way.
```

vulder commented on docs/source/vara-ts-api/plots.rst:
```diff
@@ -3,50 +3,81 @@ Plots
+4. *Plot- or table-specific options*: options specific to a certain plot
+
+Plot generation is handeled by the :class:`~varats.plot.plots.PlotGenerator` classes.
+The CLI selects a concrete plot generator via the passed **plot generator name** and immediately calls its ``__call__()`` function (which is implemented by the :class:`~varats.plot.plots.PlotGenerator` itself and must not be overridden by subclasses).
```
Should we decorate the method with `@final`?

boehmseb replied:
I wasn't aware that such a decorator exists. And yes, for now, we could definitely add that decorator.

I can imagine that in the future there maybe might be a case where we would need to override this function (e.g., for some obscure plotting framework), but if so, we can still remove the annotation.

vulder reviewed (approved):
Looks good to me.
The only thing I did not fully understand, regarding the configs, do we automatically provide different defaults depending on whether we generate a plot or present it (view) to the user?

boehmseb committed:
Update docs/source/tutorials/visualizing_data.rst

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

boehmseb committed:
Allow separate defaults for view mode in plot config

boehmseb committed:
Add @final to PlotGenerator call operator

boehmseb committed:
Fix @final import pre python 3.8

boehmseb committed:
Add missing class docstring

boehmseb committed:
Migrate artefacts extra tests to integration tests
